File: BaseConverter.py
# convert to binary
def binary(deca):
    cont = True
    result = ""
    # while deca / 2 != 0, add deca modulo 2 to the string result
    while cont:
        result = "".join((result, str(deca % 2)))
        deca = int(deca / 2)
        if deca / 2 == 0:
            cont = False
    result = result[::-1]   # reverse string (MSB to LSB)
    print("Binary: ", result)


# convert to octal
def octal(deca):
    cont = True
    result = ""
    # while deca / 8 != 0, add deca modulo 8 to the string result
    while cont:
        result = "".join((result, str(deca % 8)))
        deca = int(deca / 8)
        if deca / 8 == 0:
            cont = False
    result = result[::-1]   # reverse string (MSB to LSB)
    print("Octal: ", result)


# convert to hexadecimal
def hexadecimal(deca):
    cont = True
    result = ""
    # while deca / 16 != 0, add deca modulo 16 to the string result
    while cont:
        if deca % 16 == 10:
            current = "A"
        elif deca % 16 == 11:
            current = "B"
        elif deca % 16 == 12:
            current = "C"
        elif deca % 16 == 13:
            current = "D"
        elif deca % 16 == 14:
            current = "E"
        elif deca % 16 == 15:
            current = "F"
        else:
            current = str(deca % 16)

        result += current # append the string result with the current modulo

        deca = int(deca / 16) # new value of deca

        if deca / 16 == 0: # when to stop
            cont = False

    result = result[::-1]   # reverse string result

    print("Hexadecimal: ", result)

# user input (a decimal number that we wish to convert)
decimal = input("Enter a number base 10: ")

# print binary
binary(int(decimal))

# print octal
octal(int(decimal))

# print decimal
print("Decimal: ", decimal)

# print hexadecimal
hexadecimal(int(decimal))

File: BaseConverter.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>


// function to reverse a string
void reverse (char *input);
void reverse(char *input){
	char output[10000]="";
	for (int i = 0; i < strlen(input); i++){
		output[i] = input[strlen(input)-1-i];
	}
	printf("%s\n", output);
}




int main (int argc, char **argv){

	// take the input as a string (the input is a decimal)
	char *initialValue = argv[1];

	// convert it to an integer (still a decimal)
	int value = atoi(initialValue);



	// convert from decimal to binary
	int value1 = value;
	char binaryStr[10000];
	char currentBinary[10];
	while (value1!=0){
		sprintf(currentBinary, "%d", value1%2); //convert the current digit to a string
		strcat(binaryStr,currentBinary); // add that string digit to the binaryStr string
		value1 /= 2;
	}
	printf("Binary: ");
	reverse(binaryStr); // reverse the string to have the final result



	// convert from decimal to octal
	int value2 = value;
	char octalStr[10000];
	char currentOctal[10];
	while (value2!=0){
		sprintf(currentOctal, "%d", value2%8);
		strcat(octalStr,currentOctal);
		value2 /= 8;
	}
	printf("Octal: ");
	reverse(octalStr);



	// print the decimal version (the input)
	printf("Decimal: %s\n", initialValue);



	// convert from decimal to hex
	int value3 = value;
	char hexStr[10000];
	while(value3!=0){
		char digit[1000] = "";
		int current = value3%16;
		if (current == 10){ // check for each case from 10 to 15
			strcat(hexStr,"A"); // append hexStr with "A"
		} else if (current == 11){
			strcat(hexStr, "B");
		} else if (current == 12){
			strcat(hexStr, "C");
		} else if (current == 13){
			strcat(hexStr, "D");
		} else if (current == 14){
			strcat(hexStr, "E");
		} else if (current == 15){
			strcat(hexStr, "F");
		} else {		// if the remainder is less than 10, convert that remainder to a string and add it to the  string hexStr
			sprintf(digit, "%d", current);
			strcat(hexStr,digit);
		}
		value3/=16;
	}
	printf("Hexadecimal: 0x");
	reverse(hexStr);

	return 0;
}

